src/lib.rs

```diff
@@ -43,72 +51,89 @@ impl<T: 'static + Send + Sync> ParVec<T> {
 
     /// Take the inner `Vec` if there are no slices remaining.
     /// Returns `Err(self)` if there are still slices out there.
-    pub fn into_inner_opt(self) -> Result<Vec<T>, ParVec<T>> {
-        // Unwrap if we hold a unique reference
-        // (we don't use weak refs so ignore those)
-        if arc::strong_count(&self.data) == 1 {
-            let vec_ptr: &mut Vec<T> = unsafe { mem::transmute(&*self.data) };
-            Ok(mem::replace(vec_ptr, Vec::new()))
-        } else {
-            Err(self)
-        }
+    pub fn try_unwrap(mut self) -> Result<Vec<T>, ParVec<T>> {
+        // Unwrap if we hold a unique reference.
+        // The return is required because `self` would still be borrowed mutably in `else`.
+        if let Some(data) = arc::get_mut(&mut self.data) {
+             return Ok(mem::replace(&mut data.0, Vec::new()));
+        } 
+        
+        Err(self)
     }
 
-    /// Take the inner `Vec`, waiting until all slices have been freed.
-    pub fn into_inner(mut self) -> Vec<T> {
+    /// Take the inner `Vec`, waiting in a spinlock until all slices have been freed.
+    ///
+    /// ###Deadlock Warning
+    /// Before calling this method, you should ensure that all `ParSlice` instances have either been:
+    ///
+    /// - moved to other threads that will quit sometime in the future, or;
+    /// - dropped, implicitly (left in an inner scope) or explicitly (passed to `mem::drop()`)
+    ///
+    /// Otherwise, a deadlock will likely occur.
+    pub fn unwrap(mut self) -> Vec<T> {
         loop {
-            match self.into_inner_opt() {
+            match self.try_unwrap() {
                 Ok(vec) => return vec,
                 Err(new_self) => self = new_self,
             }
+            // Yield from the spinlock so we don't eat up too much CPU time.
+            ::std::thread::yield_now();
```

Comment:
A whole context switch per iteration probably isn't too good for performance. Maybe try doing this every X cycles of the loop?

Alternatively, use a `Condvar` and `Mutex` to manage blocking instead of a spinlock.